tensorpack/dataflow/common.py

```diff
@@ -471,3 +471,49 @@ def add_next():
                     for v in self.q:
                         yield v
                     return
+
+
+class DebugData(ProxyDataFlow):
+    """ Behave like an identity mapping but print shapes of produced datapoints once during construction."""
+
+    def reason_input_data(self, t, k, depth=1):
+        if isinstance(t, list):
+            print("%s is list of %i elements " (" " * (depth * 2), len(t)))
+        else:
+            t_shape = "n.A."
+            if hasattr(t, 'shape'):
+                t_shape = t.shape
+            t_type = t.__class__.__name__
+
+            t_max, t_min = None, None
+            if hasattr(t, 'max'):
+                t_max = t.max()
+            if hasattr(t, 'min'):
+                t_min = t.min()
+
+            t_range = ("None-None")
+            if t_max is not None or t_min is not None:
+                t_range = "%.4f-%.4f" % (t_min, t_max)
+
+            print("%s dp %i: is %s of shape %s with range [%s]" % (" " * (depth * 2), k, t_type, t_shape, t_range))
+
+    def __init__(self, ds):
+        super(DebugData, self).__init__(ds)
+        self.ds = ds
+
+        ds.reset_state()
+        dummy = next(ds.get_data())
+        if isinstance(dummy, list):
+            logger.info("Incoming dataflow consists of %i elements" % len(dummy))
+            for k, entry in enumerate(dummy):
+                self.reason_input_data(entry, k)
```

Comment:
Could you move the implementation to `reset_state()`? `reset_state` of an underlying DataFlow only get called when you call reset_state from a high level DataFlow, this is currently true for all DataFlow except Prefetch.

I'm not sure will this cause any issues, though. I'm trying to think if I'm using this assumption somewhere.